src/libstd/sys/wasm_cloudabi/condvar.rs

```diff
@@ -0,0 +1,43 @@
+// Copyright 2017 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+use sys::mutex::Mutex;
+use time::Duration;
+
+pub struct Condvar { }
+
+impl Condvar {
+    pub const fn new() -> Condvar {
+        Condvar { }
+    }
+
+    #[inline]
+    pub unsafe fn init(&mut self) {}
+
+    #[inline]
+    pub unsafe fn notify_one(&self) {
+    }
+
+    #[inline]
+    pub unsafe fn notify_all(&self) {
+    }
+
+    pub unsafe fn wait(&self, _mutex: &Mutex) {
+        panic!("can't block with web assembly")
+    }
+
+    pub unsafe fn wait_timeout(&self, _mutex: &Mutex, _dur: Duration) -> bool {
+        panic!("can't block with web assembly");
```

Comment:
(Note to reader: I am the author of CloudABI, but I know very little about wasm. Would love to know more about it, though)

Even though I understand that in Javascript land the event model is different from, say, traditional synchronous C code, is it really the case that wasm + CloudABI couldn't use the same synchronisation primitives? Wasm is just a fictive (interpreted) CPU architecture that could in principle use the same ways of doing synchronisation, right?

Reply:
If we are talking about primitives for multi-threading: there is WebAssembly threads proposal at https://github.com/WebAssembly/threads/ (There is also https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Atomics on JS side)

See "wasm_cloudabi/condvar_atomics.rs" for implemetation when atomics are enabled for wasm32- target (I shall probably link these two to files in "sys/wasm")

Reply:
> Are we certain that nothing can be done to leverage that work to wasm+Rust+CloudABI?

The scope of this PR is to set up the ground for "wasm32-unknown-cloudabi"-like target. So far, I don't see anything that stop us to use CloudABI locking primitives. There are no runtimes to verify our implementation though, so my idea is to slowly build up support and sync work with potential runtimes, e.g. [wasmtime](https://github.com/CraneStation/wasmtime).